File: app/javascript/crypto/ratchet.js
/**
 * Double Ratchet Implementation - Signal Protocol Compatible
 *
 * This implements the Double Ratchet algorithm matching Signal's specification:
 * - HKDF-SHA256 for key derivation
 * - X25519 for Diffie-Hellman ratchet
 * - AES-256-CBC + HMAC-SHA256 for authenticated encryption
 *
 * Reference: https://signal.org/docs/specifications/doubleratchet/
 */

import { x25519 } from "@noble/curves/ed25519";
import { hkdf } from "@noble/hashes/hkdf";
import { sha256 } from "@noble/hashes/sha256";
import { hmac } from "@noble/hashes/hmac";
import { cbc } from "@noble/ciphers/aes";
import { randomBytes } from "@noble/ciphers/webcrypto";

/**
 * Double Ratchet session for E2E encryption with forward secrecy.
 *
 * Each message uses a unique key derived from the ratchet state.
 * Compromising one key doesn't compromise past or future messages.
 */
export class RatchetSession {
  /**
   * Create a new RatchetSession from initial shared secret.
   *
   * @param {Uint8Array} sharedSecret - 32-byte shared secret from initial key exchange
   * @param {boolean} isInitiator - true if this party initiated the session
   */
  constructor(sharedSecret, isInitiator) {
    // Initialize root key from shared secret using HKDF
    const initial = this.kdf(sharedSecret, new Uint8Array(32), "ratchet-init", 64);
    this.rootKey = initial.slice(0, 32);

    // Generate initial DH keypair
    this.dhPrivateKey = randomBytes(32);
    this.dhPublicKey = x25519.getPublicKey(this.dhPrivateKey);

    // Peer's DH public key (set when receiving first message)
    this.peerPublicKey = null;

    // Chain keys for sending and receiving
    this.sendChainKey = null;
    this.recvChainKey = null;

    // Message counters for replay protection
    this.sendCount = 0;
    this.recvCount = 0;

    // Track previous chain length for header
    this.prevChainLength = 0;

    // Is this party the initiator? Determines initial ratchet direction
    this.isInitiator = isInitiator;

    // If initiator, derive initial send chain from root key
    if (isInitiator) {
      const chainInit = this.kdf(this.rootKey, new Uint8Array(32), "chain-init", 64);
      this.rootKey = chainInit.slice(0, 32);
      this.sendChainKey = chainInit.slice(32, 64);
    }
  }

  /**
   * HKDF-SHA256 key derivation function.
   *
   * @param {Uint8Array} inputKey - Input key material
   * @param {Uint8Array} salt - Salt (can be zeros)
   * @param {string} info - Context string
   * @param {number} length - Output length in bytes
   * @returns {Uint8Array} Derived key material
   */
  kdf(inputKey, salt, info, length = 64) {
    return hkdf(sha256, inputKey, salt, info, length);
  }

  /**
   * Advance the sending chain and return a message key.
   *
   * @returns {Uint8Array} 32-byte message key for encryption
   */
  advanceSendChain() {
    if (!this.sendChainKey) {
      throw new Error("Send chain not initialized - need DH ratchet first");
    }

    // Derive new chain key and message key
    const output = this.kdf(this.sendChainKey, new Uint8Array(32), "chain", 64);
    this.sendChainKey = output.slice(0, 32);
    const messageKey = output.slice(32, 64);

    this.sendCount++;
    return messageKey;
  }

  /**
   * Advance the receiving chain and return a message key.
   *
   * @returns {Uint8Array} 32-byte message key for decryption
   */
  advanceRecvChain() {
    if (!this.recvChainKey) {
      throw new Error("Receive chain not initialized - need DH ratchet first");
    }

    // Derive new chain key and message key
    const output = this.kdf(this.recvChainKey, new Uint8Array(32), "chain", 64);
    this.recvChainKey = output.slice(0, 32);
    const messageKey = output.slice(32, 64);

    this.recvCount++;
    return messageKey;
  }

  /**
   * Perform a DH ratchet step when receiving a new public key.
   *
   * This provides the "asymmetric" part of the Double Ratchet:
   * - Compute new shared secret from DH
   * - Derive new root key and receiving chain
   * - Generate new DH keypair
   * - Derive new sending chain
   *
   * @param {Uint8Array} peerPublicKey - Peer's new DH public key
   */
  dhRatchet(peerPublicKey) {
    this.peerPublicKey = peerPublicKey;

    // DH with current private key → derive receiving chain
    const dh1 = x25519.getSharedSecret(this.dhPrivateKey, peerPublicKey);
    const output1 = this.kdf(dh1, this.rootKey, "ratchet", 64);
    this.rootKey = output1.slice(0, 32);
    this.recvChainKey = output1.slice(32, 64);
    this.recvCount = 0;

    // Save previous chain length for header
    this.prevChainLength = this.sendCount;

    // Generate new DH keypair
    this.dhPrivateKey = randomBytes(32);
    this.dhPublicKey = x25519.getPublicKey(this.dhPrivateKey);

    // DH with new private key → derive sending chain
    const dh2 = x25519.getSharedSecret(this.dhPrivateKey, peerPublicKey);
    const output2 = this.kdf(dh2, this.rootKey, "ratchet", 64);
    this.rootKey = output2.slice(0, 32);
    this.sendChainKey = output2.slice(32, 64);
    this.sendCount = 0;
  }

  /**
   * PKCS7 padding for AES-CBC.
   *
   * @param {Uint8Array} data - Data to pad
   * @returns {Uint8Array} Padded data (multiple of 16 bytes)
   */
  pkcs7Pad(data) {
    const padLen = 16 - (data.length % 16);
    const padded = new Uint8Array(data.length + padLen);
    padded.set(data);
    padded.fill(padLen, data.length);
    return padded;
  }

  /**
   * Remove PKCS7 padding.
   *
   * @param {Uint8Array} data - Padded data
   * @returns {Uint8Array} Unpadded data
   */
  pkcs7Unpad(data) {
    const padLen = data[data.length - 1];
    if (padLen < 1 || padLen > 16) {
      throw new Error("Invalid PKCS7 padding");
    }
    // Verify padding bytes
    for (let i = data.length - padLen; i < data.length; i++) {
      if (data[i] !== padLen) {
        throw new Error("Invalid PKCS7 padding");
      }
    }
    return data.slice(0, data.length - padLen);
  }

  /**
   * Encrypt a message using the Double Ratchet.
   *
   * @param {Uint8Array} plaintext - Message to encrypt
   * @returns {Object} Encrypted envelope with header
   */
  encrypt(plaintext) {
    // Get message key from symmetric ratchet
    const messageKey = this.advanceSendChain();

    // Derive encryption key, MAC key, and IV from message key
    const derived = this.kdf(messageKey, new Uint8Array(32), "message", 80);
    const encKey = derived.slice(0, 32);
    const macKey = derived.slice(32, 64);
    const iv = derived.slice(64, 80);

    // AES-256-CBC encrypt with PKCS7 padding
    const cipher = cbc(encKey, iv);
    const padded = this.pkcs7Pad(plaintext);
    const ciphertext = cipher.encrypt(padded);

    // HMAC-SHA256 authenticate (truncated to 8 bytes like Signal)
    const macInput = new Uint8Array(ciphertext.length + 32);
    macInput.set(this.dhPublicKey);
    macInput.set(ciphertext, 32);
    const fullMac = hmac(sha256, macKey, macInput);
    const mac = fullMac.slice(0, 8);

    return {
      version: 2,
      header: {
        dhPublicKey: this.dhPublicKey,
        prevChainLength: this.prevChainLength,
        messageNumber: this.sendCount - 1,
      },
      ciphertext,
      mac,
    };
  }

  /**
   * Decrypt a message using the Double Ratchet.
   *
   * @param {Object} envelope - Encrypted envelope from encrypt()
   * @returns {Uint8Array} Decrypted plaintext
   */
  decrypt(envelope) {
    const { header, ciphertext, mac } = envelope;

    // Check if we need to perform a DH ratchet
    // (peer's DH key is different from what we have)
    const needsRatchet = !this.peerPublicKey ||
      !this.arraysEqual(header.dhPublicKey, this.peerPublicKey);

    if (needsRatchet) {
      this.dhRatchet(header.dhPublicKey);
    }

    // Get message key from symmetric ratchet
    const messageKey = this.advanceRecvChain();

    // Derive encryption key, MAC key, and IV from message key
    const derived = this.kdf(messageKey, new Uint8Array(32), "message", 80);
    const encKey = derived.slice(0, 32);
    const macKey = derived.slice(32, 64);
    const iv = derived.slice(64, 80);

    // Verify HMAC first
    const macInput = new Uint8Array(ciphertext.length + 32);
    macInput.set(header.dhPublicKey);
    macInput.set(ciphertext, 32);
    const expectedMac = hmac(sha256, macKey, macInput).slice(0, 8);

    if (!this.arraysEqual(mac, expectedMac)) {
      throw new Error("MAC verification failed - message tampered or wrong key");
    }

    // AES-256-CBC decrypt and remove padding
    const cipher = cbc(encKey, iv);
    const padded = cipher.decrypt(ciphertext);
    return this.pkcs7Unpad(padded);
  }

  /**
   * Compare two Uint8Arrays for equality.
   *
   * @param {Uint8Array} a
   * @param {Uint8Array} b
   * @returns {boolean}
   */
  arraysEqual(a, b) {
    if (a.length !== b.length) return false;
    for (let i = 0; i < a.length; i++) {
      if (a[i] !== b[i]) return false;
    }
    return true;
  }

  /**
   * Get current DH public key for inclusion in message header.
   *
   * @returns {Uint8Array} Current 32-byte DH public key
   */
  getPublicKey() {
    return this.dhPublicKey;
  }

  /**
   * Zero out all sensitive key material.
   * Call this when disconnecting.
   */
  zeroize() {
    if (this.rootKey) this.rootKey.fill(0);
    if (this.sendChainKey) this.sendChainKey.fill(0);
    if (this.recvChainKey) this.recvChainKey.fill(0);
    if (this.dhPrivateKey) this.dhPrivateKey.fill(0);

    this.rootKey = null;
    this.sendChainKey = null;
    this.recvChainKey = null;
    this.dhPrivateKey = null;
    this.dhPublicKey = null;
    this.peerPublicKey = null;
  }
}

/**
 * Serialize envelope for transmission.
 *
 * @param {Object} envelope - Envelope from RatchetSession.encrypt()
 * @returns {Object} JSON-serializable envelope
 */
export function serializeEnvelope(envelope) {
  return {
    version: envelope.version,
    header: {
      dh_public_key: bytesToBase64(envelope.header.dhPublicKey),
      prev_chain_length: envelope.header.prevChainLength,
      message_number: envelope.header.messageNumber,
    },
    ciphertext: bytesToBase64(envelope.ciphertext),
    mac: bytesToBase64(envelope.mac),
  };
}

/**
 * Deserialize envelope from transmission.
 *
 * @param {Object} data - Received envelope data
 * @returns {Object} Envelope for RatchetSession.decrypt()
 */
export function deserializeEnvelope(data) {
  return {
    version: data.version,
    header: {
      dhPublicKey: base64ToBytes(data.header.dh_public_key),
      prevChainLength: data.header.prev_chain_length,
      messageNumber: data.header.message_number,
    },
    ciphertext: base64ToBytes(data.ciphertext),
    mac: base64ToBytes(data.mac),
  };
}

// Base64 helpers
function bytesToBase64(bytes) {
  return btoa(String.fromCharCode(...bytes));
}

function base64ToBytes(str) {
  return new Uint8Array(atob(str).split("").map(c => c.charCodeAt(0)));
}

File: cli/src/relay/ratchet.rs
//! Double Ratchet Implementation - Signal Protocol Compatible
//!
//! This implements the Double Ratchet algorithm matching Signal's specification:
//! - HKDF-SHA256 for key derivation
//! - X25519 for Diffie-Hellman ratchet
//! - AES-256-CBC + HMAC-SHA256 for authenticated encryption
//!
//! Reference: https://signal.org/docs/specifications/doubleratchet/
//!
//! Rust guideline compliant 2025-01

use aes::cipher::{block_padding::Pkcs7, BlockDecryptMut, BlockEncryptMut, KeyIvInit};
use anyhow::{Context, Result};
use base64::{engine::general_purpose::STANDARD as BASE64, Engine};
use hkdf::Hkdf;
use hmac::{Hmac, Mac};
use rand::RngCore;
use serde::{Deserialize, Serialize};
use sha2::Sha256;
use zeroize::{Zeroize, ZeroizeOnDrop};

type Aes256CbcEnc = cbc::Encryptor<aes::Aes256>;
type Aes256CbcDec = cbc::Decryptor<aes::Aes256>;
type HmacSha256 = Hmac<Sha256>;

/// Double Ratchet message header
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RatchetHeader {
    /// Sender's current DH public key
    pub dh_public_key: String,
    /// Number of messages in previous sending chain
    pub prev_chain_length: u64,
    /// Message number in current sending chain
    pub message_number: u64,
}

/// Encrypted envelope with Double Ratchet header (protocol v2)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RatchetEnvelope {
    /// Protocol version (2 for Double Ratchet)
    pub version: u8,
    /// Ratchet header with DH public key
    pub header: RatchetHeader,
    /// Base64-encoded ciphertext
    pub ciphertext: String,
    /// Base64-encoded MAC (8 bytes)
    pub mac: String,
}

/// Double Ratchet session for E2E encryption with forward secrecy.
///
/// Each message uses a unique key derived from the ratchet state.
/// Compromising one key doesn't compromise past or future messages.
#[derive(Zeroize, ZeroizeOnDrop)]
pub struct RatchetSession {
    /// Root key for deriving chain keys
    root_key: [u8; 32],
    /// Current sending chain key
    send_chain_key: Option<[u8; 32]>,
    /// Current receiving chain key
    recv_chain_key: Option<[u8; 32]>,
    /// Our DH private key
    dh_private_key: [u8; 32],
    /// Our DH public key (not sensitive, but included for convenience)
    #[zeroize(skip)]
    dh_public_key: [u8; 32],
    /// Peer's DH public key
    #[zeroize(skip)]
    peer_public_key: Option<[u8; 32]>,
    /// Send message counter
    #[zeroize(skip)]
    send_count: u64,
    /// Receive message counter
    #[zeroize(skip)]
    recv_count: u64,
    /// Previous chain length (for header)
    #[zeroize(skip)]
    prev_chain_length: u64,
    /// Is this party the initiator?
    #[zeroize(skip)]
    is_initiator: bool,
}

impl std::fmt::Debug for RatchetSession {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("RatchetSession")
            .field("send_count", &self.send_count)
            .field("recv_count", &self.recv_count)
            .field("is_initiator", &self.is_initiator)
            .finish_non_exhaustive()
    }
}

impl RatchetSession {
    /// Create a new RatchetSession from initial shared secret.
    ///
    /// # Arguments
    /// * `shared_secret` - 32-byte shared secret from X25519 key exchange
    /// * `is_initiator` - true if this party initiated the session (CLI is initiator)
    pub fn new(shared_secret: &[u8; 32], is_initiator: bool) -> Result<Self> {
        // Derive initial root key using HKDF
        let initial = Self::kdf(shared_secret, &[0u8; 32], b"ratchet-init", 64)?;

        // Generate initial DH keypair
        let mut dh_private_key = [0u8; 32];
        rand::thread_rng().fill_bytes(&mut dh_private_key);
        let dh_public_key = x25519_dalek::PublicKey::from(
            &x25519_dalek::StaticSecret::from(dh_private_key)
        ).to_bytes();

        let mut root_key = [0u8; 32];
        root_key.copy_from_slice(&initial[..32]);

        let mut session = Self {
            root_key,
            send_chain_key: None,
            recv_chain_key: None,
            dh_private_key,
            dh_public_key,
            peer_public_key: None,
            send_count: 0,
            recv_count: 0,
            prev_chain_length: 0,
            is_initiator,
        };

        // Derive initial chain key from root key
        // Both parties use the same derivation so they get matching keys
        let chain_init = Self::kdf(&session.root_key, &[0u8; 32], b"chain-init", 64)?;
        session.root_key.copy_from_slice(&chain_init[..32]);

        let mut chain_key = [0u8; 32];
        chain_key.copy_from_slice(&chain_init[32..64]);

        if is_initiator {
            // Initiator uses this as their initial sending chain
            session.send_chain_key = Some(chain_key);
        } else {
            // Non-initiator uses this as their initial receiving chain
            // (to decrypt the first message from initiator)
            session.recv_chain_key = Some(chain_key);
        }

        Ok(session)
    }

    /// HKDF-SHA256 key derivation function
    fn kdf(input_key: &[u8], salt: &[u8], info: &[u8], length: usize) -> Result<Vec<u8>> {
        let hk = Hkdf::<Sha256>::new(Some(salt), input_key);
        let mut output = vec![0u8; length];
        hk.expand(info, &mut output)
            .map_err(|e| anyhow::anyhow!("HKDF expansion failed: {}", e))?;
        Ok(output)
    }

    /// Advance the sending chain and return a message key
    fn advance_send_chain(&mut self) -> Result<[u8; 32]> {
        let chain_key = self.send_chain_key
            .as_ref()
            .context("Send chain not initialized - need DH ratchet first")?;

        let output = Self::kdf(chain_key, &[0u8; 32], b"chain", 64)?;

        let mut new_chain = [0u8; 32];
        new_chain.copy_from_slice(&output[..32]);
        self.send_chain_key = Some(new_chain);

        let mut message_key = [0u8; 32];
        message_key.copy_from_slice(&output[32..64]);

        self.send_count += 1;
        Ok(message_key)
    }

    /// Advance the receiving chain and return a message key
    fn advance_recv_chain(&mut self) -> Result<[u8; 32]> {
        let chain_key = self.recv_chain_key
            .as_ref()
            .context("Receive chain not initialized - need DH ratchet first")?;

        let output = Self::kdf(chain_key, &[0u8; 32], b"chain", 64)?;

        let mut new_chain = [0u8; 32];
        new_chain.copy_from_slice(&output[..32]);
        self.recv_chain_key = Some(new_chain);

        let mut message_key = [0u8; 32];
        message_key.copy_from_slice(&output[32..64]);

        self.recv_count += 1;
        Ok(message_key)
    }

    /// Perform a DH ratchet step when receiving a new public key
    fn dh_ratchet(&mut self, peer_public_key: &[u8; 32]) -> Result<()> {
        self.peer_public_key = Some(*peer_public_key);

        // DH with current private key → derive receiving chain
        let peer_public = x25519_dalek::PublicKey::from(*peer_public_key);
        let our_secret = x25519_dalek::StaticSecret::from(self.dh_private_key);
        let dh1 = our_secret.diffie_hellman(&peer_public).to_bytes();

        let output1 = Self::kdf(&dh1, &self.root_key, b"ratchet", 64)?;
        self.root_key.copy_from_slice(&output1[..32]);
        let mut recv_chain = [0u8; 32];
        recv_chain.copy_from_slice(&output1[32..64]);
        self.recv_chain_key = Some(recv_chain);
        self.recv_count = 0;

        // Save previous chain length for header
        self.prev_chain_length = self.send_count;

        // Generate new DH keypair
        rand::thread_rng().fill_bytes(&mut self.dh_private_key);
        self.dh_public_key = x25519_dalek::PublicKey::from(
            &x25519_dalek::StaticSecret::from(self.dh_private_key)
        ).to_bytes();

        // DH with new private key → derive sending chain
        let new_secret = x25519_dalek::StaticSecret::from(self.dh_private_key);
        let dh2 = new_secret.diffie_hellman(&peer_public).to_bytes();

        let output2 = Self::kdf(&dh2, &self.root_key, b"ratchet", 64)?;
        self.root_key.copy_from_slice(&output2[..32]);
        let mut send_chain = [0u8; 32];
        send_chain.copy_from_slice(&output2[32..64]);
        self.send_chain_key = Some(send_chain);
        self.send_count = 0;

        Ok(())
    }

    /// Encrypt a message using the Double Ratchet
    pub fn encrypt(&mut self, plaintext: &[u8]) -> Result<RatchetEnvelope> {
        let message_key = self.advance_send_chain()?;

        // Derive encryption key, MAC key, and IV from message key
        let derived = Self::kdf(&message_key, &[0u8; 32], b"message", 80)?;
        let enc_key: [u8; 32] = derived[..32].try_into().expect("32 bytes");
        let mac_key: [u8; 32] = derived[32..64].try_into().expect("32 bytes");
        let iv: [u8; 16] = derived[64..80].try_into().expect("16 bytes");

        // AES-256-CBC encrypt with PKCS7 padding
        let cipher = Aes256CbcEnc::new(&enc_key.into(), &iv.into());
        let ciphertext = cipher.encrypt_padded_vec_mut::<Pkcs7>(plaintext);

        // HMAC-SHA256 authenticate (truncated to 8 bytes like Signal)
        let mut mac_input = Vec::with_capacity(32 + ciphertext.len());
        mac_input.extend_from_slice(&self.dh_public_key);
        mac_input.extend_from_slice(&ciphertext);

        let mut hmac = HmacSha256::new_from_slice(&mac_key)
            .expect("HMAC accepts any key size");
        hmac.update(&mac_input);
        let mac = hmac.finalize().into_bytes();

        Ok(RatchetEnvelope {
            version: 2,
            header: RatchetHeader {
                dh_public_key: BASE64.encode(self.dh_public_key),
                prev_chain_length: self.prev_chain_length,
                message_number: self.send_count - 1,
            },
            ciphertext: BASE64.encode(&ciphertext),
            mac: BASE64.encode(&mac[..8]), // Truncated MAC
        })
    }

    /// Decrypt a message using the Double Ratchet
    pub fn decrypt(&mut self, envelope: &RatchetEnvelope) -> Result<Vec<u8>> {
        // Decode header's DH public key
        let peer_dh_bytes = BASE64.decode(&envelope.header.dh_public_key)
            .context("Invalid DH public key encoding")?;
        let peer_dh: [u8; 32] = peer_dh_bytes.try_into()
            .map_err(|_| anyhow::anyhow!("Invalid DH public key length"))?;

        // Check if we need to perform a DH ratchet
        let needs_ratchet = self.peer_public_key
            .map(|pk| pk != peer_dh)
            .unwrap_or(true);

        if needs_ratchet {
            // Special case: non-initiator receiving first message
            // We already have recv_chain from initialization, don't override it
            // But we do need to set up the send_chain for replying
            if !self.is_initiator && self.peer_public_key.is_none() && self.recv_chain_key.is_some() {
                self.peer_public_key = Some(peer_dh);

                // Generate new DH keypair for sending
                rand::thread_rng().fill_bytes(&mut self.dh_private_key);
                self.dh_public_key = x25519_dalek::PublicKey::from(
                    &x25519_dalek::StaticSecret::from(self.dh_private_key)
                ).to_bytes();

                // Derive send chain using DH with our new keypair
                let peer_public = x25519_dalek::PublicKey::from(peer_dh);
                let our_secret = x25519_dalek::StaticSecret::from(self.dh_private_key);
                let dh = our_secret.diffie_hellman(&peer_public).to_bytes();

                let output = Self::kdf(&dh, &self.root_key, b"ratchet", 64)?;
                self.root_key.copy_from_slice(&output[..32]);
                let mut send_chain = [0u8; 32];
                send_chain.copy_from_slice(&output[32..64]);
                self.send_chain_key = Some(send_chain);
                self.send_count = 0;
            } else {
                self.dh_ratchet(&peer_dh)?;
            }
        }

        let message_key = self.advance_recv_chain()?;

        // Derive encryption key, MAC key, and IV from message key
        let derived = Self::kdf(&message_key, &[0u8; 32], b"message", 80)?;
        let enc_key: [u8; 32] = derived[..32].try_into().expect("32 bytes");
        let mac_key: [u8; 32] = derived[32..64].try_into().expect("32 bytes");
        let iv: [u8; 16] = derived[64..80].try_into().expect("16 bytes");

        // Decode ciphertext and MAC
        let ciphertext = BASE64.decode(&envelope.ciphertext)
            .context("Invalid ciphertext encoding")?;
        let mac_bytes = BASE64.decode(&envelope.mac)
            .context("Invalid MAC encoding")?;

        // Verify HMAC first
        let mut mac_input = Vec::with_capacity(32 + ciphertext.len());
        mac_input.extend_from_slice(&peer_dh);
        mac_input.extend_from_slice(&ciphertext);

        let mut hmac = HmacSha256::new_from_slice(&mac_key)
            .expect("HMAC accepts any key size");
        hmac.update(&mac_input);
        let expected_mac = hmac.finalize().into_bytes();

        // Constant-time comparison for first 8 bytes
        if mac_bytes.len() != 8 || expected_mac[..8] != mac_bytes[..] {
            anyhow::bail!("MAC verification failed - message tampered or wrong key");
        }

        // AES-256-CBC decrypt and remove padding
        let cipher = Aes256CbcDec::new(&enc_key.into(), &iv.into());
        let plaintext = cipher.decrypt_padded_vec_mut::<Pkcs7>(&ciphertext)
            .map_err(|e| anyhow::anyhow!("Decryption failed: {}", e))?;

        Ok(plaintext)
    }

    /// Get current DH public key for inclusion in message header
    pub fn public_key(&self) -> &[u8; 32] {
        &self.dh_public_key
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_ratchet_roundtrip() {
        // Simulate initial key exchange
        let shared_secret = [0x42u8; 32];

        // CLI is initiator, browser is not
        let mut cli_session = RatchetSession::new(&shared_secret, true).unwrap();
        let mut browser_session = RatchetSession::new(&shared_secret, false).unwrap();

        // CLI sends first message (as initiator, can send immediately)
        let plaintext = b"Hello from CLI!";
        let envelope = cli_session.encrypt(plaintext).unwrap();

        // Browser receives and decrypts
        let decrypted = browser_session.decrypt(&envelope).unwrap();
        assert_eq!(decrypted, plaintext);

        // Browser sends reply
        let reply = b"Hello from browser!";
        let reply_envelope = browser_session.encrypt(reply).unwrap();

        // CLI receives and decrypts
        let decrypted_reply = cli_session.decrypt(&reply_envelope).unwrap();
        assert_eq!(decrypted_reply, reply);
    }

    #[test]
    fn test_forward_secrecy() {
        let shared_secret = [0x42u8; 32];

        let mut cli = RatchetSession::new(&shared_secret, true).unwrap();
        let mut browser = RatchetSession::new(&shared_secret, false).unwrap();

        // Send multiple messages and verify each has different keys
        // (implicitly tested by the ratchet advancing)
        for i in 0..10 {
            let msg = format!("Message {}", i);
            let envelope = cli.encrypt(msg.as_bytes()).unwrap();
            let decrypted = browser.decrypt(&envelope).unwrap();
            assert_eq!(decrypted, msg.as_bytes());
        }

        // Send replies
        for i in 0..5 {
            let msg = format!("Reply {}", i);
            let envelope = browser.encrypt(msg.as_bytes()).unwrap();
            let decrypted = cli.decrypt(&envelope).unwrap();
            assert_eq!(decrypted, msg.as_bytes());
        }
    }
}
